vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java

```diff
@@ -390,4 +448,151 @@ Collection<Component> getDirectChildren() {
+            int sectionItemNodeId = (int) sectionWidgetsFromClient
+                    .getObject(sectionWidgetIdx).getNumber("nodeid");
+            section.add(nodeIdToWidget.get(sectionItemNodeId));
+        }
+    }
+
+    private boolean areItemsInSync(JsonArray itemsFromClient) {
+        // Assert root level item count
+        if (itemsFromClient.length() != getChildren().count()) {
+            return false;
+        }
+        Set<Integer> allNodeIds = new HashSet<>();
+        for (int rootLevelItemIdx = 0; rootLevelItemIdx < itemsFromClient
+                .length(); rootLevelItemIdx++) {
+            JsonObject rootLevelItemFromClient = itemsFromClient
+                    .getObject(rootLevelItemIdx);
+            int rootLevelItemNodeId = (int) rootLevelItemFromClient
+                    .getNumber("nodeid");
+            // Assert unique node ID
+            if (allNodeIds.contains(rootLevelItemNodeId)) {
+                return false;
+            }
+            allNodeIds.add(rootLevelItemNodeId);
+            Optional<Component> childMatch = getChildren().filter(child -> child
+                    .getElement().getNode().getId() == rootLevelItemNodeId)
+                    .findAny();
+            // Assert root level item node ID
+            if (childMatch.isEmpty()) {
+                return false;
+            }
+            // Assert section contents
+            if (childMatch.get() instanceof DashboardSection sectionMatch
+                    && !areSectionItemsInSync(sectionMatch,
+                            rootLevelItemFromClient, allNodeIds)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean areSectionItemsInSync(DashboardSection section,
+            JsonObject sectionFromClient, Set<Integer> allNodeIds) {
+        // Assert section has items array
+        if (!sectionFromClient.hasKey("items")) {
+            return false;
+        }
+        JsonArray sectionItemsFromClient = sectionFromClient.getArray("items");
+        // Assert section child item count
+        if (sectionItemsFromClient.length() != section.getWidgets().size()) {
+            return false;
+        }
+        for (int sectionItemIdx = 0; sectionItemIdx < sectionItemsFromClient
+                .length(); sectionItemIdx++) {
+            int sectionItemNodeId = (int) sectionItemsFromClient
+                    .getObject(sectionItemIdx).getNumber("nodeid");
+            // Assert unique node ID
+            if (allNodeIds.contains(sectionItemNodeId)) {
+                return false;
+            }
+            allNodeIds.add(sectionItemNodeId);
+            // Assert section child item node ID
+            if (section.getWidgets().stream().noneMatch(child -> child
+                    .getElement().getNode().getId() == sectionItemNodeId)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void customizeItemReorderEndEvent() {
```

Comment:
This needs to run on attach, otherwise it wouldn't work if you for example detach and reattach the dashboard:
```java
add(new NativeButton("Toggle attached", e -> {
        if (dashboard.getParent().isPresent()) {
                dashboard.removeFromParent();
        } else {
                add(dashboard);
        }
}));
```

Reply:
Fixed the issue with dashboard detach&reattach and added tests.